src/drivers/intel/cavs/ipc.c

```diff
@@ -211,6 +213,18 @@ void ipc_platform_send_msg(struct ipc *ipc)
 
 	list_item_append(&msg->list, &ipc->shared_ctx->empty_list);
 
+	/* Is this notification IPC? */
+	rhdr = (struct sof_ipc_reply *)msg->tx_data;
+	if (rhdr->hdr.cmd & SOF_IPC_COMP_NOTIFICATION) {
+		/* Dispatch registered "On Send" events */
+		event_id = ((struct sof_ipc_comp_event *)
+			   msg->tx_data)->event_type;
+		if (ipc->ipc_cb[event_id].cb) {
+			ipc->ipc_cb[event_id].cb(ipc->ipc_cb[event_id].cb_data);
```

Comment:
Hi Marcin, can you move this invoking of cb to the IPC reply handling, which is much reasonable, here:
```
/* test code to check working IRQ */
static void ipc_irq_handler(void *arg)
{
	struct ipc_msg *msg;
	uint32_t dipcctl;

...

		/* unmask Done interrupt */
		ipc_write(IPC_DIPCCTL,
			  ipc_read(IPC_DIPCCTL) | IPC_DIPCCTL_IPCIDIE);

+		msg = _ipc->shared_ctx->dsp_msg;
+
+		/* Call registered callbacks */
+		if (msg && msg->cb) {
+			msg->cb(msg->cb_data);
+			/* Clear callback afterward */
+			msg->cb = NULL;
+		}
...		
		/* send next message to host */
		ipc_process_msg_queue();
	}
}
```

Reply:
@keyonjie, IPC reply handling may look like right choice at first but in fact it will slightly delay draining. Especially now when we have large host buffer on our disposal we should start draining as soon as possible.